### PR TITLE
Randomness AIP

### PR DESCRIPTION
**Pull request by danielxiangzl**

### CODE REVIEW

File: aips/on-chain-randomness-impl.md [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+### Disaster Scenarios
+
+**DKG loses liveness**. In case of implementation bugs/bad configurations, DKG may never finish, so the epoch change won't finish and buffered on-chain config updates won't be applied. Note that the chain is still live in this case, but the validators cannot advance to new epochs.  
+
+*Mitigation*: Validators can force epoch change and disable randomness feature. The Aptos Labs need to deploy hotfix for full mitigation after debugging.
+
+**Randomness generation loses liveness**. If this happens, the blockchain will halt as every block needs randomness to proceed for execution.
+
+*Mitigation*: Validators can disable randomness feature using local configuration overwrite and restart all validators to pause the randomness generation to bring the chain back alive as a temporary mitigation. The Aptos Labs need to deploy hotfix for full mitigation after debugging.
```
Comment (danielxiangzl) [inline aips/on-chain-randomness-impl.md:356]: @zjma add more details if necessary

File: aips/on-chain-randomness-impl.md [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+```
+max_execution_gas=920
+max_io_gas=1_000
+typical_octa_per_gas=100
+max_storage_fee=200_000_000
+max_number_of_gas_units=2_000_000
+```
+Therefore, for a txn that directly or indirectly call randomness API, the required deposit is 2_000_000 octas, i.e., 2APT.
+
+TODO: add / move details about undergasing attack prevent once implementation is finalized
```
Comment (danielxiangzl) [inline aips/on-chain-randomness-impl.md:398]: cc @zjma

Comment (zjma) [inline aips/on-chain-randomness-impl.md:398]:
```suggestion
```

File: aips/on-chain-randomness-impl.md [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+See the reference implementation [here](https://github.com/aptos-labs/aptos-core/blob/1de391c3589cf2d07cb423a73c4a2a6caa299ebf/aptos-move/aptos-vm/src/validator_txns/dkg.rs#L70).
+
+
+## Testing (Optional)
+
+Unit tests, smoke tests, forge tests.
+
+Deploy a test network called `randomnet` which lasts for several months.
+
+Further stress / performance testing in `previewnet`.
```
Comment (zjma) [inline aips/on-chain-randomness-impl.md:344]:
```suggestion
Further stress / performance testing done in `previewnet`.
```

**Review by zjma — APPROVED** [review]
thank you!

File: aips/on-chain-randomness-impl.md (hunk 2 above)
Comment (zjma) [reply aips/on-chain-randomness-impl.md:398]: based on the discussion today, we probably proceed with the current implementation.

File: aips/on-chain-randomness-impl.md (hunk 1 above)
Comment (zjma) [reply aips/on-chain-randomness-impl.md:356]: only implementation details are remaining. and i think they don't need to be in the AIP.

File: aips/on-chain-randomness-impl.md [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+created: 02/12/2024
+updated (*optional): <mm/dd/yyyy>
+requires (*optional): <AIP number(s)>
+---
+
+# AIP-X - Implementation of instant on-chain randomness
+
+## Summary
+
+This AIP describes the implementation of the randomness API from AIP-41[^aip-41], focusing on the cryptographic constructions and their efficient integration into our system. Specifically, we describe how we use a weighted distributed key generation (DKG) protocol in a PoS setting to set up threshold keys among the validators, and have validators generate randomness every block using a weighted verifiable unpredictable function (VUF). We will also describe other required changes such as reconfiguration change.
```
Comment (thepomeranian) [inline aips/on-chain-randomness-impl.md:18]: this summary currently describes how we are doing the implementation. can we add a few lines about the goals and motivation behind this aip in the summary?

File: aips/on-chain-randomness-impl.md [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+
+### Goals
+
+The goal is to provide Move smart contracts with access to (1) **instant**[^aip-41], (2) **unbiasable** and (3) **unpredictable** randomness w.r.t. to any minority stake of the validators.
```
Comment (thepomeranian) [inline aips/on-chain-randomness-impl.md:22]: remove shorthand `w.r.t`

File: aips/on-chain-randomness-impl.md [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+
+At the same time, we must do this without imposing a high penalty on the blockchain’s latency or throughput.
+
+## Motivation
+
+As already described in AIP-41[^aip-41] the need for on-chain randomness arises in many applications, such as decentralized games, raffles, randomized NFTs, randomized airdrops, and more. Plus, many blockchain applications will benefit from randomness in terms of fairness, security, and functionality.
+
+## Impact
+
+On-chain randomness interacts with and impacts our blockchain protocol. The two protocols must be carefully co-designed to maintain performance, especially in terms of latency.
```
Comment (thepomeranian) [inline aips/on-chain-randomness-impl.md:32]: does this aip impact the indexer or any other apps?

File: aips/on-chain-randomness-impl.md [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+
+**Interactive VSS and VUFs**. A different design would use a DKG based on interactive VSS to deal a secret amongst our validators. Unfortunately, this design requires that new-epoch validators be online when setting up their threshold keys for randomness generation, which does not align with our current system architecture.
+
+**Other insecure approaches.** Several existing blockchains have insecure on-chain randomness implementations, as described in our [blog post](https://aptoslabs.medium.com/roll-with-move-secure-instant-randomness-on-aptos-c0e219df3fb1). 
+
+## Specification
+
+On-chain randomness introduces several changes to the current Aptos blockchain system, including running a DKG for every epoch change, modifying the existing reconfiguration process, and generating a **randomness seed** for every block. This section describes the above system changes at a high level. Implementation details are discussed in [Reference Implementation](#reference-implementation).
```
Comment (thepomeranian) [inline aips/on-chain-randomness-impl.md:44]: can we add in who is running or who can run the dkg? is this a service hosted by aptos labs?

File: aips/on-chain-randomness-impl.md (hunk 4 above)
Comment (danielxiangzl) [reply aips/on-chain-randomness-impl.md]: done

File: aips/on-chain-randomness-impl.md (hunk 5 above)
Comment (danielxiangzl) [reply aips/on-chain-randomness-impl.md]: done

File: aips/on-chain-randomness-impl.md (hunk 6 above)
Comment (danielxiangzl) [reply aips/on-chain-randomness-impl.md]: done

File: aips/on-chain-randomness-impl.md (hunk 7 above)
Comment (danielxiangzl) [reply aips/on-chain-randomness-impl.md]: done

**Review by sherry-x — APPROVED** [review]

**Review by thepomeranian — APPROVED** [review]

**Pull request merged**